meta/rpc.go

```diff
@@ -149,6 +112,63 @@ func (r *rpc) handleRPCConn(conn net.Conn) {
 	r.sendResponse(conn, typ, resp)
 }
 
+func (r *rpc) readMessage(conn net.Conn) (internal.RPCType, []byte, error) {
+	// Read request size.
+	var sz uint64
+	if err := binary.Read(conn, binary.BigEndian, &sz); err != nil {
+		return internal.RPCType_Error, nil, fmt.Errorf("read size: %s", err)
+	}
+
+	if sz == 0 {
+		return internal.RPCType_Error, nil, fmt.Errorf("invalid message size: %d", sz)
+	}
+
+	if sz >= MaxMessageSize {
+		return internal.RPCType_Error, nil, fmt.Errorf("max message size of %d exceeded: %d", MaxMessageSize, sz)
```

Comment:
@jwilder this originally returned `0` which didn't seem valid, so I updated it to `internal.RPCType_Error`